crates/kumod/src/smtp_server.rs

```diff
@@ -2264,6 +2280,177 @@ impl SmtpServerSession {
+                        }
+                    }
+                }
+                _ = self.shutdown.shutting_down() => {
+                    return Err(anyhow!("shutting down"));
+                }
+            };
+            let header = HeaderResult::parse(&self.read_buffer);
+            if header.is_complete() {
+                break header;
+            }
+        };
+
+        let mut addr: Option<IpAddr> = None;
+        let mut port: Option<u16> = None;
+        let mut dest_addr: Option<IpAddr> = None;
+        let mut dest_port: Option<u16> = None;
+        match header {
+            HeaderResult::V1(Ok(header)) => {
+                let addresses = header.addresses;
+                match addresses {
+                    ppp::v1::Addresses::Tcp4(addresses) => {
+                        addr.replace(addresses.source_address.into());
+                        port.replace(addresses.source_port);
+                        dest_addr.replace(addresses.destination_address.into());
+                        dest_port.replace(addresses.destination_port);
+                    }
+                    ppp::v1::Addresses::Tcp6(addresses) => {
+                        addr.replace(addresses.source_address.into());
+                        port.replace(addresses.source_port);
+                        dest_addr.replace(addresses.destination_address.into());
+                        dest_port.replace(addresses.destination_port);
+                    }
+                    _ => {}
+                }
+            }
+            HeaderResult::V2(Ok(header)) => {
+                let addresses = header.addresses;
+                match addresses {
+                    ppp::v2::Addresses::IPv4(addresses) => {
+                        addr.replace(addresses.source_address.into());
+                        port.replace(addresses.source_port);
+                        dest_addr.replace(addresses.destination_address.into());
+                        dest_port.replace(addresses.destination_port);
+                    }
+                    ppp::v2::Addresses::IPv6(addresses) => {
+                        addr.replace(addresses.source_address.into());
+                        port.replace(addresses.source_port);
+                        dest_addr.replace(addresses.destination_address.into());
+                        dest_port.replace(addresses.destination_port);
+                    }
+                    _ => {}
+                }
+            }
+            HeaderResult::V1(Err(error)) => {
+                return Err(anyhow!("proxy protocol v1 parsing error: {error}"))
+            }
+            HeaderResult::V2(Err(error)) => {
+                return Err(anyhow!("proxy protocol v2 parsing error: {error}"))
+            }
+        }
+        self.read_buffer.clear();
+
+        let old_peer = self.peer_address.clone();
+        let old_via = self.my_address.clone();
+
+        self.peer_address = (addr.unwrap(), port.unwrap()).into();
```

Comment:
These unwraps are undesirable because an attacker could send a proxy protocol packet with an Unknown or Unspecified address to trigger a panic and break the service.

We need to handle this situation gracefully!  Using `SocketAddr` (as indicated above) to have fewer pieces to consider here makes it a bit less verbose to conditionally update the pieces here using `if let Some(..) = ..`